adserver/decisionengine/backends.py

```diff
@@ -404,15 +404,15 @@ def select_flight(self):
                     flight.niche_targeting
                     and "ethicalads_ext.embedding" in settings.INSTALLED_APPS
                 ):
-                    if not self.niche_weights:
-                        from ethicalads_ext.embedding.utils import (  # noqa
-                            get_niche_weights,
-                        )
+                    # We have to do this here,
+                    # so we can filter by the weight in the filter_flight call below
+                    # TODO: Handle filtering by multiple flights in the future
+                    from ethicalads_ext.embedding.utils import (  # noqa
+                        get_niche_weights,
+                    )
 
-                        self.niche_weights = get_niche_weights(
-                            self.url,
-                        )
-                        log.info("Niche targeting weights: %s", self.niche_weights)
+                    self.niche_weights = get_niche_weights(self.url, flights=[flight])
```

Comment:
This may be fine for a test but this will quickly fail to scale and might even just be wrong as written if there's 2 niche targeting flights. This method (`get_niche_weights`) currently makes 1-2 queries per call with 2 on success. I assume this is what your comment above is about. However, it seems it would be pretty easy to do something like the following at the top of the method:


```python
flights_with_niche_targeting = [flight.niche_targeting for flight in flights]
if any(flights_with_niche_targeting):
    self.niche_weights = get_niche_weights(self.url, flights=flights_with_niche_targeting)
```

Am I missing something?

Reply:
I was mostly just trying to test out the filtering. This will work, but just do a query for each flight with it enabled. I'm fine aggregating the flights upfront, it shouldn't be hard as you mentioned.

Reply:
I did see an issue with what I wrote above. If we have niche targeting only in house flights, it will do the aggregation even if we're selecting a paid flight since it will look at all flights. We could do this on a flight-type basis. Regardless, I think what you have is fine for a test but need to be very mindful of scaling.

Reply:
Yea, the different types of flights and priority makes it a bit odd. We could do the basic aggregation on flight type if we wanted, as a middle ground to start.

Reply:
> We could do the basic aggregation on flight type if we wanted, as a middle ground to start.

I think we should do that (if I'm understanding correctly). I don't want a niche targeted house flight to slow down paid ads but I'm ok if it slows down some house ads.

Reply:
@davidfischer This will already not run on paid ads, if the only ads with niche_targeting are house ads.

Reply:
I understand that but when we centralize the code to make it more performant, we have to ensure that stays true.